File: morse-translator-zig/src/morse.zig
// ============================================================================
//  Morse-Pi — Morse code translation, encoding / decoding, word lists.
// ============================================================================
const std = @import("std");
const state = @import("state.zig");

// ── Morse code lookup tables ─────────────────────────────────────────────────

pub const Entry = struct { char: u8, code: []const u8 };

pub const MORSE_TABLE = [_]Entry{
    .{ .char = 'A', .code = ".-" },
    .{ .char = 'B', .code = "-..." },
    .{ .char = 'C', .code = "-.-." },
    .{ .char = 'D', .code = "-.." },
    .{ .char = 'E', .code = "." },
    .{ .char = 'F', .code = "..-." },
    .{ .char = 'G', .code = "--." },
    .{ .char = 'H', .code = "...." },
    .{ .char = 'I', .code = ".." },
    .{ .char = 'J', .code = ".---" },
    .{ .char = 'K', .code = "-.-" },
    .{ .char = 'L', .code = ".-.." },
    .{ .char = 'M', .code = "--" },
    .{ .char = 'N', .code = "-." },
    .{ .char = 'O', .code = "---" },
    .{ .char = 'P', .code = ".--." },
    .{ .char = 'Q', .code = "--.-" },
    .{ .char = 'R', .code = ".-." },
    .{ .char = 'S', .code = "..." },
    .{ .char = 'T', .code = "-" },
    .{ .char = 'U', .code = "..-" },
    .{ .char = 'V', .code = "...-" },
    .{ .char = 'W', .code = ".--" },
    .{ .char = 'X', .code = "-..-" },
    .{ .char = 'Y', .code = "-.--" },
    .{ .char = 'Z', .code = "--.." },
    .{ .char = '0', .code = "-----" },
    .{ .char = '1', .code = ".----" },
    .{ .char = '2', .code = "..---" },
    .{ .char = '3', .code = "...--" },
    .{ .char = '4', .code = "....-" },
    .{ .char = '5', .code = "....." },
    .{ .char = '6', .code = "-...." },
    .{ .char = '7', .code = "--..." },
    .{ .char = '8', .code = "---.." },
    .{ .char = '9', .code = "----." },
    .{ .char = '.', .code = ".-.-.-" },
    .{ .char = ',', .code = "--..--" },
    .{ .char = '?', .code = "..--.." },
    .{ .char = '\'', .code = ".----." },
    .{ .char = '!', .code = "-.-.--" },
    .{ .char = '/', .code = "-..-." },
    .{ .char = '(', .code = "-.--." },
    .{ .char = ')', .code = "-.--.-" },
    .{ .char = '&', .code = ".-..." },
    .{ .char = ':', .code = "---..." },
    .{ .char = ';', .code = "-.-.-." },
    .{ .char = '=', .code = "-...-" },
    .{ .char = '+', .code = ".-.-." },
    .{ .char = '-', .code = "-....-" },
    .{ .char = '_', .code = "..--.-" },
    .{ .char = '"', .code = ".-..-." },
    .{ .char = '$', .code = "...-..-" },
    .{ .char = '@', .code = ".--.-." },
};

/// Look up Morse code for a character.
pub fn charToMorse(c: u8) ?[]const u8 {
    const upper = std.ascii.toUpper(c);
    if (upper == ' ') return "/";
    for (MORSE_TABLE) |e| {
        if (e.char == upper) return e.code;
    }
    return null;
}

/// Look up character for a Morse code string.
pub fn morseToChar(code: []const u8) u8 {
    if (std.mem.eql(u8, code, "/")) return ' ';
    for (MORSE_TABLE) |e| {
        if (std.mem.eql(u8, e.code, code)) return e.char;
    }
    return '?';
}

/// Encode plain text to Morse code string.
pub fn encode(text: []const u8) []u8 {
    var buf = std.ArrayList(u8).init(state.gpa);
    for (text) |c| {
        if (charToMorse(c)) |code| {
            if (buf.items.len > 0) buf.append(' ') catch {};
            buf.appendSlice(code) catch {};
        }
    }
    return buf.toOwnedSlice() catch &.{};
}

/// Decode Morse code string to plain text.
pub fn decode(morse_str: []const u8) []u8 {
    var buf = std.ArrayList(u8).init(state.gpa);
    var it = std.mem.splitScalar(u8, morse_str, ' ');
    while (it.next()) |code| {
        if (code.len == 0) continue;
        buf.append(morseToChar(code)) catch {};
    }
    return buf.toOwnedSlice() catch &.{};
}

/// Write the Morse code lookup table as JSON (matches Python MORSE_CODE dict).
pub fn writeMorseTableJson(w: anytype) !void {
    try w.writeAll("{");
    var first = true;
    for (MORSE_TABLE) |e| {
        if (!first) try w.writeAll(",");
        first = false;
        try std.fmt.format(w, "\"{c}\":\"{s}\"", .{ e.char, e.code });
    }
    try w.writeAll(",\" \":\"/\"}");
}

// ── Word lists ───────────────────────────────────────────────────────────────

const WordCategory = struct {
    name: []const u8,
    words: []const []const u8,
};

// Fallback word lists (overridden if words.json is loaded)
const fallback_articles = [_][]const u8{ "the", "a", "an" };
const fallback_adjectives = [_][]const u8{ "big", "small", "red", "blue", "happy", "sad", "fast", "slow" };
const fallback_nouns = [_][]const u8{ "dog", "cat", "house", "car", "apple", "book", "tree", "river" };
const fallback_verbs = [_][]const u8{ "run", "jump", "eat", "sleep", "read", "write", "think" };
const fallback_adverbs = [_][]const u8{ "quickly", "slowly", "loudly", "quietly" };

var words_articles: []const []const u8 = &fallback_articles;
var words_adjectives: []const []const u8 = &fallback_adjectives;
var words_nouns: []const []const u8 = &fallback_nouns;
var words_verbs: []const []const u8 = &fallback_verbs;
var words_adverbs: []const []const u8 = &fallback_adverbs;

pub fn loadWords() void {
    const data = std.fs.cwd().readFileAlloc(state.gpa, "words.json", 1 << 20) catch return;
    defer state.gpa.free(data);

    const parsed = std.json.parseFromSlice(std.json.Value, state.gpa, data, .{}) catch return;
    defer parsed.deinit();

    if (parsed.value != .object) return;
    const obj = parsed.value.object;

    inline for (.{
        .{ "articles", &words_articles },
        .{ "adjectives", &words_adjectives },
        .{ "nouns", &words_nouns },
        .{ "verbs", &words_verbs },
        .{ "adverbs", &words_adverbs },
    }) |pair| {
        if (obj.get(pair[0])) |arr| {
            if (arr == .array) {
                var list = std.ArrayList([]const u8).init(state.gpa);
                for (arr.array.items) |item| {
                    if (item == .string) {
                        list.append(state.gpa.dupe(u8, item.string) catch continue) catch {};
                    }
                }
                pair[1].* = list.toOwnedSlice() catch pair[1].*;
            }
        }
    }
}

var prng: std.Random.DefaultPrng = std.Random.DefaultPrng.init(0);
var prng_seeded: bool = false;

fn getRandom() std.Random {
    if (!prng_seeded) {
        prng = std.Random.DefaultPrng.init(@as(u64, @bitCast(std.time.milliTimestamp())));
        prng_seeded = true;
    }
    return prng.random();
}

fn pickRandom(list: []const []const u8) []const u8 {
    if (list.len == 0) return "TEST";
    return list[getRandom().intRangeAtMost(usize, 0, list.len - 1)];
}

/// Generate a random phrase based on difficulty.
pub fn randomPhrase(difficulty: []const u8) []u8 {
    var buf = std.ArrayList(u8).init(state.gpa);

    if (std.mem.eql(u8, difficulty, "easy")) {
        // Single noun or verb
        const all = state.gpa.alloc([]const u8, words_nouns.len + words_verbs.len) catch
            return state.gpa.dupe(u8, "TEST") catch &.{};
        @memcpy(all[0..words_nouns.len], words_nouns);
        @memcpy(all[words_nouns.len..], words_verbs);
        const picked = pickRandom(all);
        state.gpa.free(all);
        for (picked) |c| buf.append(std.ascii.toUpper(c)) catch {};
    } else if (std.mem.eql(u8, difficulty, "hard")) {
        // article + adjective + noun + verb + adverb
        const parts = [_][]const []const u8{
            words_articles,
            words_adjectives,
            words_nouns,
            words_verbs,
            words_adverbs,
        };
        for (parts, 0..) |list, i| {
            if (list.len == 0) continue;
            if (i > 0 and buf.items.len > 0) buf.append(' ') catch {};
            for (pickRandom(list)) |c| buf.append(std.ascii.toUpper(c)) catch {};
        }
    } else {
        // medium: adjective + noun
        if (words_adjectives.len > 0) {
            for (pickRandom(words_adjectives)) |c| buf.append(std.ascii.toUpper(c)) catch {};
        }
        if (words_nouns.len > 0) {
            if (buf.items.len > 0) buf.append(' ') catch {};
            for (pickRandom(words_nouns)) |c| buf.append(std.ascii.toUpper(c)) catch {};
        }
    }

    if (buf.items.len == 0) buf.appendSlice("THE QUICK FOX") catch {};
    return buf.toOwnedSlice() catch state.gpa.dupe(u8, "TEST") catch &.{};
}

File: morse-translator-rust/src/morse.rs
// ============================================================================
//  Morse-Pi — Morse code translation, encoding / decoding, word lists.
// ============================================================================
use rand::seq::SliceRandom;
use rand::thread_rng;
use std::fs;

// ── Morse code lookup table ──────────────────────────────────────────────────

pub static MORSE_TABLE: &[(char, &str)] = &[
    ('A', ".-"),    ('B', "-..."),  ('C', "-.-."),  ('D', "-.."),
    ('E', "."),     ('F', "..-."),  ('G', "--."),   ('H', "...."),
    ('I', ".."),    ('J', ".---"),  ('K', "-.-"),   ('L', ".-.."),
    ('M', "--"),    ('N', "-."),    ('O', "---"),    ('P', ".--."),
    ('Q', "--.-"),  ('R', ".-."),   ('S', "..."),    ('T', "-"),
    ('U', "..-"),   ('V', "...-"),  ('W', ".--"),    ('X', "-..-"),
    ('Y', "-.--"),  ('Z', "--.."),
    ('0', "-----"), ('1', ".----"), ('2', "..---"),  ('3', "...--"),
    ('4', "....-"), ('5', "....."), ('6', "-...."),  ('7', "--..."),
    ('8', "---.."), ('9', "----."),
    ('.', ".-.-.-"), (',', "--..--"), ('?', "..--.."), ('\'', ".----."),
    ('!', "-.-.--"), ('/', "-..-."), ('(', "-.--."),  (')', "-.--.-"),
    ('&', ".-..."),  (':', "---..."), (';', "-.-.-."), ('=', "-...-"),
    ('+', ".-.-."),  ('-', "-....-"), ('_', "..--.-"), ('"', ".-..-."),
    ('$', "...-..-"), ('@', ".--.-."),
];

/// Look up Morse code for a character.
pub fn char_to_morse(c: char) -> Option<&'static str> {
    let upper = c.to_ascii_uppercase();
    if upper == ' ' {
        return Some("/");
    }
    MORSE_TABLE.iter().find(|(ch, _)| *ch == upper).map(|(_, code)| *code)
}

/// Look up character for a Morse code string.
pub fn morse_to_char(code: &str) -> char {
    if code == "/" {
        return ' ';
    }
    MORSE_TABLE
        .iter()
        .find(|(_, c)| *c == code)
        .map(|(ch, _)| *ch)
        .unwrap_or('?')
}

/// Encode plain text to Morse code string.
pub fn encode(text: &str) -> String {
    let mut buf = String::new();
    for c in text.chars() {
        if let Some(code) = char_to_morse(c) {
            if !buf.is_empty() {
                buf.push(' ');
            }
            buf.push_str(code);
        }
    }
    buf
}

/// Decode Morse code string to plain text.
pub fn decode(morse_str: &str) -> String {
    let mut buf = String::new();
    for code in morse_str.split(' ') {
        if code.is_empty() {
            continue;
        }
        buf.push(morse_to_char(code));
    }
    buf
}

/// Write the Morse code lookup table as JSON string.
pub fn morse_table_json() -> String {
    let mut entries: Vec<String> = MORSE_TABLE
        .iter()
        .map(|(ch, code)| format!("\"{}\":\"{}\"", ch, code))
        .collect();
    entries.push("\" \":\"/\"".into());
    format!("{{{}}}", entries.join(","))
}

// ── Word lists ───────────────────────────────────────────────────────────────

pub struct WordLists {
    pub articles: Vec<String>,
    pub adjectives: Vec<String>,
    pub nouns: Vec<String>,
    pub verbs: Vec<String>,
    pub adverbs: Vec<String>,
}

impl Default for WordLists {
    fn default() -> Self {
        Self {
            articles: vec!["the".into(), "a".into(), "an".into()],
            adjectives: vec![
                "big".into(), "small".into(), "red".into(), "blue".into(),
                "happy".into(), "sad".into(), "fast".into(), "slow".into(),
            ],
            nouns: vec![
                "dog".into(), "cat".into(), "house".into(), "car".into(),
                "apple".into(), "book".into(), "tree".into(), "river".into(),
            ],
            verbs: vec![
                "run".into(), "jump".into(), "eat".into(), "sleep".into(),
                "read".into(), "write".into(), "think".into(),
            ],
            adverbs: vec!["quickly".into(), "slowly".into(), "loudly".into(), "quietly".into()],
        }
    }
}

static WORDS: once_cell::sync::Lazy<std::sync::Mutex<WordLists>> =
    once_cell::sync::Lazy::new(|| std::sync::Mutex::new(WordLists::default()));

pub fn load_words() {
    let data = match fs::read_to_string("words.json") {
        Ok(d) => d,
        Err(_) => return,
    };
    let parsed: serde_json::Value = match serde_json::from_str(&data) {
        Ok(v) => v,
        Err(_) => return,
    };
    let obj = match parsed.as_object() {
        Some(o) => o,
        None => return,
    };

    let mut w = WORDS.lock().unwrap();

    fn extract_strings(obj: &serde_json::Map<String, serde_json::Value>, key: &str) -> Option<Vec<String>> {
        obj.get(key)?.as_array().map(|arr| {
            arr.iter()
                .filter_map(|v| v.as_str().map(|s| s.to_string()))
                .collect()
        })
    }

    if let Some(v) = extract_strings(obj, "articles") { w.articles = v; }
    if let Some(v) = extract_strings(obj, "adjectives") { w.adjectives = v; }
    if let Some(v) = extract_strings(obj, "nouns") { w.nouns = v; }
    if let Some(v) = extract_strings(obj, "verbs") { w.verbs = v; }
    if let Some(v) = extract_strings(obj, "adverbs") { w.adverbs = v; }
}

fn pick_random(list: &[String]) -> &str {
    let mut rng = thread_rng();
    list.choose(&mut rng).map(|s| s.as_str()).unwrap_or("TEST")
}

/// Generate a random phrase based on difficulty.
pub fn random_phrase(difficulty: &str) -> String {
    let w = WORDS.lock().unwrap();
    let mut rng = thread_rng();

    match difficulty {
        "easy" => {
            // Single noun or verb
            let mut all: Vec<&str> = Vec::new();
            for s in &w.nouns { all.push(s); }
            for s in &w.verbs { all.push(s); }
            let picked = all.choose(&mut rng).map(|s| *s).unwrap_or("TEST");
            picked.to_uppercase()
        }
        "hard" => {
            // article + adjective + noun + verb + adverb
            let parts: Vec<&[String]> = vec![
                &w.articles, &w.adjectives, &w.nouns, &w.verbs, &w.adverbs,
            ];
            let mut result = String::new();
            for list in parts {
                if list.is_empty() { continue; }
                if !result.is_empty() { result.push(' '); }
                result.push_str(&pick_random(list).to_uppercase());
            }
            if result.is_empty() { "THE QUICK FOX".into() } else { result }
        }
        _ => {
            // medium: adjective + noun
            let mut result = String::new();
            if !w.adjectives.is_empty() {
                result.push_str(&pick_random(&w.adjectives).to_uppercase());
            }
            if !w.nouns.is_empty() {
                if !result.is_empty() { result.push(' '); }
                result.push_str(&pick_random(&w.nouns).to_uppercase());
            }
            if result.is_empty() { "THE QUICK FOX".into() } else { result }
        }
    }
}
